foundry/applications/dialog.d.ts

```diff
@@ -162,10 +162,10 @@ declare class Dialog extends Application {
     content: string;
     yes?: (html: HTMLElement) => Y;
     no?: (html: HTMLElement) => N;
-    render: (html: HTMLElement) => void;
+    render?: (html: HTMLElement) => void;
     defaultYes?: boolean;
     rejectClose?: R;
-    options?: Partial<Dialog.Options> & { jQuery?: false };
+    options?: Partial<Dialog.Options> & { jQuery: false };
```

Comment:
Making this optional would allow to for the second overload to match if the user passes a callback that explicitly only handles `HTMLElement`s and no `JQuery`s if he doesn't pass the options. That would potentially cause a runtime error in his code because he will actually receive a `JQuery`. So `options` should not be optional in this overload.

The same is _sort of_ true for the 3rd overload but it doesn't matter because the function the user passes in handles both types anyways. I still suggest to make `options` non optional in the third overload anyways because if you don't pass any options, you will receive `JQuery`, so overload 1 should be the one that matches.

The same applies to `prompt`.

Reply:
Technically, the only way for it to be HTMLElement is if the user has explicitly set `options.jQuery: false`. There is no state in which the params could be `JQuery | HTMLElement`

So there should only be 2 overloads of the function (not 3). One with an optional `options` param and callbacks that return `JQuery`, and the overload where `jQuery` is explicitly `true` and has the callbacks receive an HTMLElement.

Reply:
No, there could be cases where the user passes a `boolean` that is not known at compile time as `options.jQuery`. That's what the 3rd overload is for.

Reply:
I see, so in a situation where they use something like `mergeObject` to obtain a jQuery value in an obfuscated way that is not "knowable" to the compiler.

Reply:
Now would
```
options: Partial<Dialog.Options> & { jQuery?: true }
```
be sufficient, or would we need `jQuery` to be explicit?

Reply:
I think the following would be correct:
```ts
static prompt<T>({
    title,
    content,
    label,
    callback,
    render,
    options
  }: {
    title?: string;
    content: string;
    label?: string;
    callback: (html: JQuery) => T;
    render?: (html: JQuery) => void;
    options?: Partial<Dialog.Options> & { jQuery?: true };
  }): Promise<T>;
  static prompt<T>({
    title,
    content,
    label,
    callback,
    render,
    options
  }: {
    title: string;
    content: string;
    label?: string;
    callback: (html: HTMLElement) => T;
    render?: (html: HTMLElement) => void;
    options: Partial<Dialog.Options> & { jQuery: false };
  }): Promise<T>;
  static prompt<T>({
    title,
    content,
    label,
    callback,
    render,
    options
  }: {
    title: string;
    content: string;
    label?: string;
    callback: (html: JQuery | HTMLElement) => T;
    render?: (html: JQuery | HTMLElement) => void;
    options: Partial<Dialog.Options>;
  }): Promise<T>;
```

Reply:
In theory, you can also leave out `content` and `title` and it will still "work". I just tried it, only if `callback` is missing does it actually break. But I don't think that's the intended usage.

Reply:
Also note that I removed the optionality of the complete destructured argument. While `foundry.js` _does_ provide `{}` as default argument, this will just throw an error when the button is clicked because `callback` is undefined. So I think we should not allow it.

Reply:
Haha yeah I actually just changed to that. No I don't think leaving out `content` and `title` is intended, but they also are not explicitly required. I would say they should be optional since the API does not restrict them.

Reply:
So basically the case you were describing is correct. The reason is that in this case, the type of `options.jQuery` is `true | undefined`. It can never be `false` at runtime, so we know at compile time that `JQuery` will be passed as `html`

Reply:
Or we could try to enforce it, if someone really doesn't want to have a title/content, they can just pass an explicit empty string

Reply:
I don't really mind...